UNITTESTS/moduletests/storage/kvstore/FileSystemStore/moduletest.cpp

```diff
@@ -0,0 +1,111 @@
+/* Copyright (c) 2020 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "features/storage/blockdevice/HeapBlockDevice.h"
+#include "features/storage/kvstore/filesystemstore/FileSystemStore.h"
+#include "features/storage/filesystem/littlefs/LittleFileSystem.h"
+#include "mbed_error.h"
+#include <stdlib.h>
+
+#define HEAPBLOCK_SIZE (4096)
+
+using namespace mbed;
+
+class FileSystemStoreModuleTest : public testing::Test {
+protected:
+    HeapBlockDevice heap{HEAPBLOCK_SIZE};
+    LittleFileSystem *fs;
+    FileSystemStore *store;
+
+    virtual void SetUp()
+    {
+        fs = new LittleFileSystem("kvstore", &heap);
+        if(fs->mount(&heap) != MBED_SUCCESS) {
+            EXPECT_EQ(fs->reformat(&heap), MBED_SUCCESS);
+        }
+        store = new FileSystemStore(fs);
+        EXPECT_EQ(store->init(), MBED_SUCCESS);
+    }
+
+    virtual void TearDown()
+    {
+        EXPECT_EQ(store->deinit(), MBED_SUCCESS);
+        delete store;
+        EXPECT_EQ(fs->unmount(), MBED_SUCCESS);
+        delete fs;
+    }
+};
+
+TEST_F(FileSystemStoreModuleTest, init)
+{
+    EXPECT_EQ(store->deinit(), MBED_SUCCESS);
+    EXPECT_EQ(store->init(), MBED_SUCCESS);
+    EXPECT_EQ(store->init(), MBED_SUCCESS);
+}
+
+TEST_F(FileSystemStoreModuleTest, set_get)
+{
+    char buf[100];
+    size_t size;
+    EXPECT_EQ(store->set("key", "data", 5, 0), MBED_SUCCESS);
+    EXPECT_EQ(store->get("key", buf, 100, &size), MBED_SUCCESS);
+    EXPECT_EQ(size, 5);
+    EXPECT_STREQ("data", buf);
+}
+
+TEST_F(FileSystemStoreModuleTest, erased_set_get)
+{
+    EXPECT_EQ(store->deinit(), MBED_SUCCESS);
```

Comment:
I assume these test can't be run in random order.

Reply:
I think the tests are run in the order they are written. But that shouldn't matter as the tests are separate and setup and teardown are ran for each test.

Reply:
@jarlamsa This test case fails if `BlockDevice::erase()` actually performs an erase, unless we call `fs->reformat()` afterwards. We didn't see this issue because `HeapBlockDevice::erase()` is essentially a no-op. But now I'm working on de-allocating the heap buffer on erase, and get the same failure. (Same happens, if we memset the buffer instead of de-allocate it.)

I can add the reformat to make it pass. But erasing the underlying BlockDevice sounds like restarting everything from scratch? Could you please explain me the use case we're trying to test here? Thanks in advance.

Reply:
Why do you want to deallocate the heap buffer on erase? That doesn't sound like correct behavior.

Reply:
If deallocating the heap is not correct, then what about a BlockDevice that does an actual erase (e.g. like many flash devices) and sets values to `0xFF`, for example? If I didn't miss anything, this test case relies on the assumption that data is unaltered when `BlockDevice::erase()` is called?

Reply:
Referring to this test case, I think it can be safely deleted, erasing the whole heap underneath the filesystem just corrupts the filesystem.
 
But regarding the deallocation of the allocated memory from the heap when erasing, it really sounds just wrong. Now the allocation for `HeapBlockDevice` is done on init and deallocation is done on destructor, how do you plan to change this?

Reply:
`HeapBlockDevice::init()` only allocates an array or null pointers to blocks, not individual blocks. Only when a block is used in `HeapBlockDevice::program()`, that particular block is allocated and added to the array of pointers.
My `HeapBlockDevice::erase()` [change](https://github.com/ARMmbed/mbed-os/pull/14483/commits/1a5bdad0b6734f6fe2348a02feed2e255a4db2b0) only frees blocks being erased. The destructor, as before, frees all block _and_ the array of pointers. So in terms of behaviour, it's still self-consistent, I think.

But if the change looks redundant I can remove it.

Reply:
Ok, that explains it. LGTM.

Reply:
Cheers